libraries/AP_FETtecOneWire/AP_FETtecOneWire.cpp

```diff
@@ -0,0 +1,730 @@
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/* Protocol implementation was provided by FETtec */
+/* Strongly modified by Amilcar Lucas, IAV GmbH */
+
+#include <AP_Math/AP_Math.h>
+#include <AP_Scheduler/AP_Scheduler.h>
+#include <AP_SerialManager/AP_SerialManager.h>
+#include <SRV_Channel/SRV_Channel.h>
+#include <GCS_MAVLink/GCS.h>
+
+#include "AP_FETtecOneWire.h"
+#if HAL_AP_FETTEC_ONEWIRE_ENABLED
+
+extern const AP_HAL::HAL& hal;
+
+static constexpr uint32_t TRANSACTION_TIME_US = 2000;
+static constexpr uint32_t FW_START_TIME_US = 5000;
+static constexpr uint32_t BAUDRATE = 500000;
+static constexpr uint8_t ALL_ID = 0x1F;
+static constexpr uint8_t FRAME_OVERHEAD = 6;
+static constexpr uint8_t MAX_TRANSMIT_LENGTH = 4;
+static constexpr uint8_t MAX_RECEIVE_LENGTH = 12;
+static constexpr uint8_t MAX_RESPONSE_LENGTH = FRAME_OVERHEAD + MAX_RECEIVE_LENGTH;
+
+const AP_Param::GroupInfo AP_FETtecOneWire::var_info[] = {
+    // @Param: MASK
+    // @DisplayName: Channel Bitmask
+    // @Description: Enable of FETtec OneWire ESC protocol to specific channels
+    // @Bitmask: 0:Channel1,1:Channel2,2:Channel3,3:Channel4,4:Channel5,5:Channel6,6:Channel7,7:Channel8,8:Channel9,9:Channel10,10:Channel11,11:Channel12,12:Channel13,13:Channel14,14:Channel15,15:Channel16
+    // @RebootRequired: True
+    // @User: Standard
+    AP_GROUPINFO("MASK",  1, AP_FETtecOneWire, _motor_mask, 0),
+
+    // @Param: POLES
+    // @DisplayName: Nr. electrical poles
+    // @Description: Number of motor electrical poles
+    // @Range: 2 50
+    // @RebootRequired: False
+    // @User: Standard
+    AP_GROUPINFO("POLES", 2, AP_FETtecOneWire, _pole_count, 14),
+
+    AP_GROUPEND
+};
+
+AP_FETtecOneWire *AP_FETtecOneWire::_singleton;
+
+AP_FETtecOneWire::AP_FETtecOneWire()
+{
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+    if (_singleton != nullptr) {
+        AP_HAL::panic("AP_FETtecOneWire must be singleton");
+    }
+#endif
+    _singleton = this;
+
+    _response_length[OW_OK] = 1;
+    _response_length[OW_BL_START_FW] = 0;       // BL only
+    _response_length[OW_REQ_TYPE] = 1;
+    _response_length[OW_REQ_SN] = 12;
+    _response_length[OW_REQ_SW_VER] = 2;
+    _response_length[OW_SET_FAST_COM_LENGTH] = 1;
+    _response_length[OW_SET_TLM_TYPE] = 1;
+}
+
+/**
+  initialize the serial port, scan the OneWire bus, setup the found ESCs
+*/
+void AP_FETtecOneWire::init()
+{
+    if (!_uart_initialised) {
+        AP_SerialManager& serial_manager = AP::serialmanager();

```

Comment:
I don't know if serial_manager = AP::serialmanager(); ever returns null on error, but its common practice in a lot of ardupilot code.  if so, doing a null->find_serial(...) is bad.   i'd put a null pointer check in here before calling serial_manager.find_serial